shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/rewrite/token/ProjectionsTokenGeneratorTest.java

```diff
@@ -0,0 +1,165 @@
+
+    private static final String TEST_LOGIC_TABLE_NAME = "TEST_LOGIC_TABLE_NAME";
+
+    private static final String TEST_ACTUAL_TABLE_NAME = "TEST_ACTUAL_TABLE_NAME";
+
+    private static final String TEST_ACTUAL_TABLE_NAME_WRAPPED = "TEST_ACTUAL_TABLE_NAME_WRAPPED";
+
+    private static final String TEST_OTHER_DERIVED_PROJECTION_ALIAS = "TEST_OTHER_DERIVED_PROJECTION_ALIAS";
+
+    private static final String TEST_OTHER_DERIVED_PROJECTION_EXPRESSION = "TEST_OTHER_DERIVED_PROJECTION_EXPRESSION";
+
+    private RouteUnit routeUnit = mock(RouteUnit.class);
+
+    @Before
+    public void setup() {
+        RouteMapper routeMapper = mock(RouteMapper.class);
+        when(routeMapper.getLogicName()).thenReturn(TEST_LOGIC_TABLE_NAME);
+        when(routeMapper.getActualName()).thenReturn(TEST_ACTUAL_TABLE_NAME);
+        Collection<RouteMapper> routeMapperCollection = new LinkedList<>();
+        routeMapperCollection.add(routeMapper);
+        when(routeUnit.getTableMappers()).thenReturn(routeMapperCollection);
+    }
+
+    @Test
+    public void assertIsGenerateSQLToken() {
+        ProjectionsTokenGenerator projectionsTokenGenerator = getProjectionsTokenGenerator();
+        InsertStatementContext insertStatementContext = mock(InsertStatementContext.class);
+        assertFalse(projectionsTokenGenerator.isGenerateSQLToken(insertStatementContext));
+        SelectStatementContext selectStatementContext = mock(SelectStatementContext.class, RETURNS_DEEP_STUBS);
+        when(selectStatementContext.getProjectionsContext().getProjections()).thenReturn(Collections.emptyList());
+        assertFalse(projectionsTokenGenerator.isGenerateSQLToken(selectStatementContext));
+        AggregationProjection aggregationProjection = getAggregationProjection();
+        Collection<Projection> projectionCollection = new LinkedList<>();
+        projectionCollection.add(aggregationProjection);
+        when(selectStatementContext.getProjectionsContext().getProjections()).thenReturn(projectionCollection);
+        assertTrue(projectionsTokenGenerator.isGenerateSQLToken(selectStatementContext));
+    }
+
+    @Test
+    public void assertGenerateSQLToken() {
+        Collection<Projection> projectionCollection = new LinkedList<>();
+        AggregationProjection aggregationProjection = getAggregationProjection();
+        projectionCollection.add(aggregationProjection);
+        DerivedProjection derivedProjection = getDerivedProjection();
+        projectionCollection.add(derivedProjection);
+        DerivedProjection otherDerivedProjection = getOtherDerivedProjection();
+        projectionCollection.add(otherDerivedProjection);
+        SelectStatementContext selectStatementContext = mock(SelectStatementContext.class, RETURNS_DEEP_STUBS);
+        when(selectStatementContext.getProjectionsContext().getProjections()).thenReturn(projectionCollection);
+        when(selectStatementContext.getProjectionsContext().getStopIndex()).thenReturn(TEST_STOP_INDEX);
```

Comment:
If `TEST_STOP_INDEX ` is used in only one method, please use local variable substitution.